### PR TITLE
move: minimal end-to-end source service

### PR DESCRIPTION
**Pull request by rvantonder**
## Description 

Adds end-to-end support to serve source code given an address and module, e.g.,:

```
curl http://0.0.0.0:8000/api?address=0x2&module=address
```

This is minimal feature set, and the things below will make things more feature complete. Right now I'd like to push for end-to-end integration with explorer, which will drive the remaining features.

- support `testnet`/`mainnet`/`devnet` networks via URL params (currently we use the current sui environment network)
- better file read management (cache proxy and/or DB blobstore and/or in-memory store or caching)
- support for retrieving all package modules (e.g., something like `address=0x2&module=*`). This mostly in anticipation for a variety of ways a client might decide to retrieve source.

## Test Plan 

Extends the integration test to resolve address/module pairs for an example package.

### CODE REVIEW

File: crates/sui-source-validation-service/src/lib.rs [hunk 1]
```diff
@@ -197,18 +221,59 @@ pub async fn verify_packages(
...
+    let symbol = Symbol::from(module);
+    let Ok(address) = AccountAddress::from_hex_literal(&address) else {
+	let error = format!("Invalid hex address {address}");
+	return (StatusCode::BAD_REQUEST, Json(ErrorResponse { error }).into_response())
+    };
+    let Some(path) = app_state.sources.get(&(address, symbol)) else {
+	let error = format!("No source found for {symbol} at address {address}" );
+	return (StatusCode::BAD_REQUEST, Json(ErrorResponse { error }).into_response())
+    };
+    let Ok(source) = fs::read_to_string(path) else {
```
Comment (amnn) [inline crates/sui-source-validation-service/src/lib.rs:271]: We may want to cache these in memory rather than read them from disk for every request (and in future employ a more elaborate caching mechanism) -- even though we don't have many packages, I expect people might look at the system packages fairly often.

File: crates/sui-source-validation-service/src/lib.rs [hunk 2]
```diff
@@ -197,18 +221,59 @@ pub async fn verify_packages(
...
+    Query(Request { address, module }): Query<Request>,
+) -> impl IntoResponse {
+    let symbol = Symbol::from(module);
+    let Ok(address) = AccountAddress::from_hex_literal(&address) else {
+	let error = format!("Invalid hex address {address}");
+	return (StatusCode::BAD_REQUEST, Json(ErrorResponse { error }).into_response())
+    };
+    let Some(path) = app_state.sources.get(&(address, symbol)) else {
+	let error = format!("No source found for {symbol} at address {address}" );
+	return (StatusCode::BAD_REQUEST, Json(ErrorResponse { error }).into_response())
```
Comment (amnn) [inline crates/sui-source-validation-service/src/lib.rs:269]: This seems like a 404 kind of error, not a 400.

**Review by amnn — APPROVED** [review]
Change looks good!  Some things to consider (which don't need to block anything because we're not using this in production yet):

#### Versioning
We are going to want to version things, so let's think about that from the get go.  The current state of play for RPC is to version using an HTTP header (`X-Sui-RPC-Version`).  We shouldn't use the same version for this, because it should be free to evolve separately, but could use the same versioning scheme: `X-Sui-Source-Validation-Version`?

#### Interface
Curious to get @Jordan-Mysten's take as he will be using this more than anyone else.  We should also consider if the query interface is going to evolve into something richer, or become something that we provide as part of the Explorer's extended RPC service.

If these eventualities seem likely, we should consider GraphQL sooner than later, as it will change most/(all?) of the interface. @oxade is already looking into this for RPC in general.

If we do decide to keep the interface distinct, then I do have a soft-spot for `api/[address]/[module]` over `api?address=[address]&module=[module]`, to take advantage of the hierarchy, and reduce typing, but this is a take it or leave it kind of suggestion.

**Comment by rvantonder** [comment]
Awesome call-outs, thanks @amnn!

> **Versioning** [...] could use the same versioning scheme: X-Sui-Source-Validation-Version?

Sounds good to me, will iterate on this bit, pending @Jordan-Mysten comments

> **Interface** 

I frontloaded some of the conversation and Jordan is/was in favor of REST for this service because it simplifies not dealing with more involved changes to RPC client IIRC.

> we should consider GraphQL sooner than later

The production-ready REST interface should end up being pretty simple, and convertible to GQL. The converse, successfully designing and predicting for GQL and changing the schema as needs evolves, is more tricky, in my experience. I would love if we had a firm grasp on what the GQL interface + client support in the frontend for RPC changes before going into GQL for this service. My preferred attack is:

- minimal production-ready REST interface (+ simple changes to sui explorer client based on chatting with Jordan)
- optional: iteration on REST interface depending on evolving needs
- eventual migration to GQL when there's a clear direction on plumbing / client implementation (possibly the existing RPC should migrate to GQL before I see this service doing so).

(and to be clear I'm happy to take responsibility for that)

> If we do decide to keep the interface distinct, then I do have a soft-spot for api/[address]/[module] over api?address=[address]&module=[module], to take advantage of the hierarchy, and reduce typing, but this is a take it or leave it kind of suggestion.

Thought a bit about this, two things stand out:

- Fitting in optionality for selecting the network (`network=...`). This could conceivably be added in the hierarchy like `api/[network]/[address]/[module]`, but also:

- I think introducing parameters is unavoidable if we introduce something like `module=*` or other filter mechanisms, which seem like a reasonable thing that may come up. In that case, I'm less a fan of the "impure" mixture of paths + optional parameters (i.e., that full hierarchy property won't be omnipresent and disappears, e.g., with something like `api/[address]?module=*`).

Basically, I'd prefer either a predictable, most pure paths approach, without parameters, or a pure parameters approach. Based on the utility of having parameters support non-path values like `*` or filtering, I'd prefer to stick with parameters, it really simplifies routing and deserializing.

File: crates/sui-source-validation-service/src/lib.rs (hunk 1 above)
Comment (rvantonder) [reply crates/sui-source-validation-service/src/lib.rs:271]: `+1` going to iterate on this
